integration/unplugin/src/index.ts

```diff
@@ -238,26 +233,86 @@ const civetUnplugin = createUnplugin((options: PluginOptions = {}) => {
       if (!isCivetTranspiled(id)) return null;
 
       const filename = path.resolve(process.cwd(), id.slice(0, -outExt.length));
-      const code = await fs.promises.readFile(filename, 'utf-8');
+      const rawCivetSource = await fs.promises.readFile(filename, 'utf-8');
       this.addWatchFile(filename);
 
-      // Ideally this should have been done in a `transform` step
-      // but for some reason, webpack seems to be running them in the order
-      // of `resolveId` -> `loadInclude` -> `transform` -> `load`
-      // so we have to do transformation here instead
-      const compiled = civet.compile(code, {
-        // inlineMap: true,
-        filename: id,
-        js: transpileToJS,
-        sourceMap: true,
-      });
+      let compiled: {
+        code: string;
+        sourceMap: SourceMap | string;
+      } = undefined!;
+
+      if (options.ts === 'civet' && !transformTS) {
+        compiled = civet.compile(rawCivetSource, {
+          filename: id,
+          js: true,
+          sourceMap: true,
+        });
+      } else {
+        const compiledTS = civet.compile(rawCivetSource, {
+          filename: id,
+          js: false,
+          sourceMap: true,
+        });
+
+        sourceMaps.set(
+          path.resolve(process.cwd(), id),
+          compiledTS.sourceMap as SourceMap
+        );
+
+        if (transformTS) {
+          const resolved = path.resolve(process.cwd(), id);
+          fsMap.set(resolved, compiledTS.code);
+          // Vite and Rollup normalize filenames to use `/` instead of `\`.
+          // We give the TypeScript VFS both versions just in case.
+          const slashed = slash(resolved);
+          if (resolved !== slashed) fsMap.set(slashed, rawCivetSource);
+        }
 
-      sourceMaps.set(path.resolve(process.cwd(), id), compiled.sourceMap);
+        switch (options.ts) {
+          case 'esbuild': {
+            const esbuildTransform = (await import('esbuild')).transform;
+            const result = await esbuildTransform(compiledTS.code, {
+              loader: 'tsx',
+              sourcefile: id,
+              sourcemap: 'external',
+            });
 
-      const jsonSourceMap = compiled.sourceMap.json(
-        path.basename(id.replace(/\.[jt]sx$/, '')),
-        path.basename(id)
-      );
+            compiled = { code: result.code, sourceMap: result.map };
+            break;
+          }
+          case 'tsc': {
+            const tsTranspile = (await getTS()).transpileModule;
+            const result = tsTranspile(compiledTS.code, { compilerOptions });
```

Comment:
If `transformTS` is set (e.g. typechecking is on), does this cause TypeScript to transpile *again*, or does it re-use the existing transpilation? Ideally we'd like the latter.

Reply:
With typechecking on, we only have typescript emit the declarations and not the transpiled source.
Generating declarations requires Typescript to have the entire program with imports resolved the file map, which we can only have in `buildEnd`. So this does cause Typescript to transpile _again_ as we can't reuse, but here it doesn't care about the diagnostics or the module resolution.